### PR TITLE
clusterversion,sql,backupccl: remove PublicSchemasWithDescriptors

### PR DESCRIPTION
**Pull request by celiala**
This commit removes the 22.1 version gate `PublicSchemasWithDescriptors`, which was referenced in:

- pkg/ccl/backupccl/
- pkg/ccl/changefeedccl/
- pkg/config/zonepb/zone.go
- pkg/sql/catalog/descs/
- pkg/sql/descriptor.go
- pkg/sql/importer/
- pkg/sql/rename_table.go
- pkg/sql/reparent_database.go

Cleanup was done following guidance from [21.2 cleanup](https://github.com/cockroachdb/cockroach/pull/74270#issuecomment-1000693734):

> For the most part, if the gates were just simple if !version.IsActive { return x } or something, I just removed the block, and even if it was a little more complicated, like args = [x]; if version { args = append(args, y) }; foo(args) I still tried to mostly inline it such that it looked natural (i.e. remove that append and make it args = [x, y]).

> However for just a couple more complicated cases that were referring to <21.2 versions that needed to be replaced when those were deleted, I added a placeholder clusterversion.TODOPre21_2 alias for 21.2. Replacing those calls with this alias shouldn't change their behavior -- it was already always true, since the code today should never run in a <21.2 cluster -- but means we can delete those older versions in the meantime and then the owners of these bits can decide how to update them.

Partially addresses https://github.com/cockroachdb/cockroach/issues/80663

Release note: none
Release justification: remove old version gates

### CODE REVIEW

File: pkg/sql/rename_table.go [hunk 1]
```diff
@@ -172,39 +171,11 @@ func (n *renameTableNode) startExec(params runParams) error {
...
-				columns = append(columns, *col)
-			}
-		}
-		for _, c := range columns {
-			if c.Type.UserDefined() {
-				return pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
-					"cannot change database of table if any of its column types are user-defined")
-			}
-		}
+		// After 22.1 all public schemas will be backed by a descriptor and will be a regular UDS.
```
Comment (postamar) [inline pkg/sql/rename_table.go:174]: This comment needs to be rewritten also.

File: pkg/sql/importer/import_job.go [hunk 2]
```diff
@@ -129,21 +128,19 @@ func (r *importResumer) Resume(ctx context.Context, execCtx interface{}) error {
...
-					// The Public schema will not change in the database so both the
-					// oldSchemaIDToName and newSchemaIDToName entries will be the
-					// same for the Public schema.
-					_, dbDesc, err := descsCol.GetImmutableDatabaseByID(ctx, txn, details.ParentID, tree.DatabaseLookupFlags{Required: true})
-					if err != nil {
-						return err
-					}
-					schemaMetadata.oldSchemaIDToName[dbDesc.GetSchemaID(tree.PublicSchema)] = tree.PublicSchema
-					schemaMetadata.newSchemaIDToName[dbDesc.GetSchemaID(tree.PublicSchema)] = tree.PublicSchema
+				// In 22.1, the Public schema should always be present in the database.
```
Comment (postamar) [inline pkg/sql/importer/import_job.go:131]: This comment could be rewritten as "The public schema is expected to always be present in the database".

**Review by postamar — APPROVED** [review]
LGTM, thanks for doing this. @RichardJCai can you take a look also if you have time, please ? In particular, should anything be done to the backup tests or is it safe to outright remove these such as has been done? I think it's OK, but still.

**Review by RichardJCai — APPROVED** [review]
Only thing from me is I'd like to keep the cleanup on fail backup test. Other than that LGTM

> Reviewed 17 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-:-N97HMM993qLy78zYXLo:b-4ygzxq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala, @postamar, and @RichardJCai)

___
*[`pkg/ccl/backupccl/restore_old_versions_test.go` line 1104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-N97H8UrCvauPIkpRu4T:-N97H8UrCvauPIkpRu4U:b-smsq71) ([raw file](https://github.com/cockroachdb/cockroach/blob/24906bd59e975f607e9ea88cbd4ae5ff0432df45/pkg/ccl/backupccl/restore_old_versions_test.go#L1104)):*
> ```Go
> 						Server: &server.TestingKnobs{
> 							DisableAutomaticVersionUpgrade: make(chan struct{}),
> 							BinaryVersionOverride:          clusterversion.ByKey(clusterversion.PublicSchemasWithDescriptors - 1),
> ```

I'd like to keep this test, I think other ones can go, I think this test will still pass if we remove the BinaryVersionOverride

**Review by celiala — COMMENTED** [review]
TFTRs!

> Only thing from me is I'd like to keep the cleanup on fail backup test. 

Ack:

- left tests in `restore_old_versions_test.go`
- updated comments

I'll go ahead and merge once gets pass.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-:-N9E-CGkDCENKjNpLzY9:bx9otln)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @RichardJCai)

___
*[`pkg/ccl/backupccl/restore_old_versions_test.go` line 1104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-N97H8UrCvauPIkpRu4T:-N9DysNo9_l-tJb4xmTB:b-u9z05t) ([raw file](https://github.com/cockroachdb/cockroach/blob/24906bd59e975f607e9ea88cbd4ae5ff0432df45/pkg/ccl/backupccl/restore_old_versions_test.go#L1104)):*
Ack. 

> if we remove the BinaryVersionOverride

I wasn't clear what to do for this? 

But for now I've replaced `PublicSchemasWithDescriptors` with `TODOPreV22_1` so that we can keep the test while still removing the gate.



_Code quote:_
```Go
sqlDB.CheckQueryResults(t, `SELECT x FROM test.public.t`, [][]string{{"3"}
```
___
*[`pkg/sql/rename_table.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-N92XszJlQeP1iU6M0BL-r1-174:-N9E--IQ-iBYvyuy6jh6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d1bdb44217934d11f4ebf428e8116eae462d2c9/pkg/sql/rename_table.go#L174)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This comment needs to be rewritten also.
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_job.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85825#-N92XszJlQeP1iU6M0BH-r1-131:-N9E-0UG1-sffyuLwmQs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d1bdb44217934d11f4ebf428e8116eae462d2c9/pkg/sql/importer/import_job.go#L131)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This comment could be rewritten as "The public schema is expected to always be present in the database".
</blockquote></details>

Done.

**Comment by celiala** [comment]
These tests are failing - @postamar @RichardJCai - could you weigh-in on whether these are flakes or legit? `TestRowLevelTTLJobRandomEntries/one_column_pk_multiple_nodes` seems to require a test adjustment, but I'm unsure of how? (also, feel free to take-over/add to this PR as needed, if easier :) )

**Bazel Essential CI:**
- pkg/ccl/serverccl/serverccl_test_/serverccl_test
  - exit code 142

**GitHub CI:** 
- [TestRowLevelTTLJobRandomEntries/one_column_pk_multiple_nodes](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests/6062913?showRootCauses=true&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true)
	- see details below

**Bazel Extended CI:**
- No pass/skip/fail event found for test
	- TestMetadataSST

```
------- Stdout: -------
=== RUN   TestRowLevelTTLJobRandomEntries/one_column_pk_multiple_nodes
    ttljob_test.go:546: test case: ttljob_test.testCase{desc:"one column pk multiple nodes", createTable:"CREATE TABLE tbl (\n\tid UUID PRIMARY KEY DEFAULT gen_random_uuid(),\n\ttext TEXT\n) WITH (ttl_expire_after = '30 days')", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1001, numNonExpiredRows:5, numSplits:10, forceNonMultiTenant:false, numNodes:5, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *tree.CreateTable, time.Time))(nil)}
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
    ttljob_test.go:679:
          Error Trace:  ttljob_test.go:679
          Error:        Not equal:
                        expected: 1001
                        actual  : 1002
          Test:         TestRowLevelTTLJobRandomEntries/one_column_pk_multiple_nodes
    --- FAIL: TestRowLevelTTLJobRandomEntries/one_column_pk_multiple_nodes (13.87s)
```

**Comment by celiala** [comment]
Closing in favor of https://github.com/cockroachdb/cockroach/pull/86225

**Comment by celiala** [comment]
TFTRs!

Incorporated additions from https://github.com/cockroachdb/cockroach/pull/86225 to get PR to pass - thank you! 🙏🏼 

bors r+

**Pull request merged**